mapiproxy/servers/default/nspi/dcesrv_exchange_nsp.c

```diff
@@ -246,71 +251,86 @@ static void dcesrv_NspiUnbind(struct dcesrv_call_state *dce_call,
-	uint32_t			row, row_max;
-	TALLOC_CTX			*local_mem_ctx;
-	struct PropertyTagArray_r	*mids;
 
 	DEBUG(3, ("exchange_nsp: NspiUpdateStat (0x2)"));
-
-	/* Step 0. Ensure incoming user is authenticated */
+	/* Ensure incoming user is authenticated */
 	if (!dcesrv_call_authenticated(dce_call)) {
 		DEBUG(1, ("No challenge requested by client, cannot authenticate\n"));
 		DCESRV_NSP_RETURN(r, MAPI_E_LOGON_FAILED, NULL);
 	}
 
+	if (IS_WINUNICODE_CODEPAGE(r)) {
+		DCESRV_NSP_RETURN(r, MAPI_E_NO_SUPPORT, NULL);
+	}
+
 	emsabp_ctx = dcesrv_find_emsabp_context(&r->in.handle->uuid);
 	if (!emsabp_ctx) {
 		DCESRV_NSP_RETURN(r, MAPI_E_CALL_FAILED, NULL);
 	}
 
-	local_mem_ctx = talloc_zero(NULL, TALLOC_CTX);
-
-	/* Step 1. Sanity Checks (MS-NSPI Server Processing Rules) */
+	/* Check container exists */
 	if (r->in.pStat->ContainerID && (emsabp_tdb_lookup_MId(emsabp_ctx->tdb_ctx, r->in.pStat->ContainerID) == false)) {
-		retval = MAPI_E_INVALID_BOOKMARK;
-		goto end;
+		DCESRV_NSP_RETURN(r, MAPI_E_INVALID_BOOKMARK, NULL);
 	}
 
+	/* After the checks do the update stat */
+	dcesrv_do_NspiUpdateStat(mem_ctx, r, emsabp_ctx);
+}
+
+/**
+This method does the work of NspiUpdateStat, it is separated from dcesrv_NspiUpdateStat to be 
+able to be called from NspiQueryRows without doing again the same sanity checks and reusing the emsabp_ctx
+*/
+static void dcesrv_do_NspiUpdateStat(TALLOC_CTX *mem_ctx, struct NspiUpdateStat *r, struct emsabp_context *emsabp_ctx)
+{
+	enum MAPISTATUS			retval = MAPI_E_SUCCESS;
+	enum MAPISTATUS                 ret;
+	uint32_t			row, row_max;
+	TALLOC_CTX			*local_mem_ctx;
+	struct PropertyTagArray_r	*mids;
+
+	local_mem_ctx = talloc_zero(NULL, TALLOC_CTX);
+
 	mids = talloc_zero(local_mem_ctx, struct PropertyTagArray_r);
         if (!mids) {
                 DCESRV_NSP_RETURN(r, MAPI_E_NOT_ENOUGH_MEMORY, NULL);
         }
 
-        ret = emsabp_search(local_mem_ctx, emsabp_ctx, mids, NULL, r->in.pStat, 0);
+	ret = emsabp_search(local_mem_ctx, emsabp_ctx, mids, NULL, r->in.pStat, 0);
 	if (ret != MAPI_E_SUCCESS) {
 		row_max = 0;
-                if (ret == MAPI_E_CALL_FAILED) {
-                        retval = ret;
-                        goto end;
-                }
-	}
-	else {
-		row_max = mids->cValues;
+		if (ret == MAPI_E_CALL_FAILED) {
+			retval = ret;
+			goto end;
+		}
+	} else {
+		row_max = mids->cValues - 1;
 	}
 
 	if (r->in.pStat->CurrentRec == MID_CURRENT) {
-		/* Fractional positioning (3.1.1.4.2) */
+		/* Fractional positioning (3.1.4.5.2) */
 		row = r->in.pStat->NumPos * row_max / r->in.pStat->TotalRecs;
 		if (row > row_max) {
 			row = row_max;
 		}
-	}
-	else {
+	} else {
+		/* Absolute positioning (3.1.4.5.1) */
 		if (r->in.pStat->CurrentRec == MID_BEGINNING_OF_TABLE) {
 			row = 0;
 		}
 		else if (r->in.pStat->CurrentRec == MID_END_OF_TABLE) {
 			row = row_max;
-		}
-		else {
+		} else {
 			retval = MAPI_E_NOT_FOUND;
 			row = 0;
-			while (row < row_max) {
+			while (row <= row_max) {
```

Comment:
row_max could be `mids->cValues` and `mids->aulPropTag` is `talloc_array(mem_ctx, uint32_t, mods->cValues)` so I don't understand this change.